src/alert-pcap.c

```diff
@@ -0,0 +1,406 @@
+static TmEcode AlertPcapLogThreadInit(ThreadVars *, void *, void **);
+static TmEcode AlertPcapLogThreadDeinit(ThreadVars *, void *);
+static OutputCtx *AlertPcapLogInitCtx(ConfNode *);
+static void AlertPcapLogDeInitCtx(OutputCtx *);
+
+void TmModuleAlertPcapLogRegister(void) 
+{
+    tmm_modules[TMM_ALERTPCAP].name = MODULE_NAME;
+    tmm_modules[TMM_ALERTPCAP].ThreadInit = AlertPcapLogThreadInit;
+    tmm_modules[TMM_ALERTPCAP].Func = AlertPcapLogProcess;
+    tmm_modules[TMM_ALERTPCAP].ThreadDeinit = AlertPcapLogThreadDeinit;
+    tmm_modules[TMM_ALERTPCAP].RegisterTests = NULL;
+
+    OutputRegisterModule(MODULE_NAME, "alert-pcap", AlertPcapLogInitCtx);
+    return;
+}
+
+AlertPcapFile* AlertPcapFileNew(const char *filename, const char* directory, Packet *p) {
+    
+    AlertPcapFile *output = SCMalloc(sizeof(AlertPcapFile));
+    if (unlikely(output == NULL)) {
+        SCLogError(SC_ERR_FATAL, "Fatal error could not create AlertPcapFile output.");
+        return NULL;            
+    }
+    
+    output->filename = SCCalloc(1, strlen(filename) + 1);
+    if (unlikely(output->filename == NULL)) {
+        SCLogError(SC_ERR_FATAL, "Fatal error could not create output file for AlertPcapFile.");
+        return NULL;
+    }
+
+    snprintf(output->filename, strlen(filename), "%s", filename);
+
+    struct stat stat_buf;
+    if (stat(directory, &stat_buf) != 0) {
+        int ret;
+        ret = MakePath(filename, S_IRWXU|S_IXGRP|S_IRGRP);
+        if (ret != 0) {
+            int err = errno;
+            if (err != EEXIST) {
+                SCLogError(SC_ERR_LOGDIR_CONFIG,
+                           "Cannot create file drop directory %s: %s",
+                           directory, strerror(err));
+                exit(EXIT_FAILURE);
+            }
+        } else {
+            SCLogInfo("Created alert log pcap directory %s",
+                      directory);
+        }
+    }    
+
+    output->pcap_file = pcap_open_dead(p->datalink, 65535);
+    if (output->pcap_file == NULL) {
+        SCLogError(SC_ERR_FATAL, "Fatal error, could not create TimeMachine pcap output. Exiting...");
+        exit(EXIT_FAILURE);
+    }
+
+    output->pcap_writer = pcap_dump_open(output->pcap_file, filename);
+    if (output->pcap_writer == NULL) {
+        SCLogError(SC_ERR_LOGDIR_CONFIG, 
+                   "Cannot create alert pcap log output file %s: %s",
+                   filename, pcap_geterr(output->pcap_file));
+        exit(EXIT_FAILURE);
+    }    
+
+    output->output_file=pcap_dump_file(output->pcap_writer);
+    return output;
+}
+
+void AlertPcapFileClose(AlertPcapFile* output) {
+    
+    if (output->pcap_writer) {
+        pcap_dump_flush(output->pcap_writer);
+        pcap_dump_close(output->pcap_writer);
+    }
+    
+    if (output->pcap_file) {
+        pcap_close(output->pcap_file);
+    }
+}
+
+static TmEcode AlertPcapLogProcess(ThreadVars *t, Packet *p, void *thread_data, PacketQueue *pq,
```

Comment:
Function documentation is missing.

If I understand the code correctly, this function is used to write out the pcap for a flow. It should be entered for the packet that triggered an alert, and the packets from the same flow after that, correct? Most packets will exit early?

If I am correct, it may make sense to switch API's then. OutputRegisterPacketModule() allows you to register a logger (AlertPcapLogProcess) and a condition function which controls whether we call the logger or not.